File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int checkSum(long n);

int main(void)
{
    long cc_num = 0;

    // Retrieve user input
    do
    {
        cc_num = get_long("Number: ");
    }
    while (cc_num != (long) cc_num); // While cc_num is not a long type numeral

    // Check if numbers have valid number of digits
    if (cc_num < pow(10, 12) || cc_num > pow(10, 16) - 1)
    {
        printf("INVALID\n");
        return (0);
    }

    // Checksum to validate credit card numbers
    if (checkSum(cc_num) == 0)
    {
        // Check credit card companies based on signature
        if (cc_num / (long) pow(10, 13) == 34 || cc_num / (long) pow(10, 13) == 37)
        {
            printf("AMEX\n");
        }
        else if (cc_num / (long) pow(10, 14) > 50 && cc_num / (long) pow(10, 14) < 56)
        {
            printf("MASTERCARD\n");
        }
        else if (cc_num / (long) pow(10, 12) == 4 || cc_num / (long) pow(10, 15) == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int checkSum(long n)
{
    int sum = 0;
    int i = 0;

    while (i < 16)
    {
        // Sum odd position digits from the back
        if (i % 2 == 0)
        {
            sum += n % 10;
        }
        // Multiply by 2 and add digits together for even position digits
        else
        {
            int num = (n % 10) * 2;
            sum += num % 10 + num / 10;
        }
        // Remove last digit after summing each digit
        n /= 10;
        i++;
    }
    return sum % 10;
}

File: pset6/credit/credit.py
from cs50 import get_int


def main():

    # Request user input
    num = get_int("Number: ")

    if checkSum(num):
        cardCompany(num)
    else:
        print("INVALID")

# Checksum with luhn's algorithm


def checkSum(num):

    sum = 0

    # Sum of digits with Luhn's algorithm
    for i, n in enumerate(reversed(str(num))):
        if i % 2 == 0:
            sum += int(n)
        else:
            for j in str(int(n) * 2):
                sum += int(j)

    # Check divisibility by 10
    if sum % 10 == 0:
        return True
    return False

# Function to check card company


def cardCompany(num):

    n = int(str(num)[0:2])
    l = len(str(num))
    
    # Check AMEX
    if n in [34, 37] and l == 15:
        print("AMEX")
    
    # Check MASTERCARD
    elif n in range(51, 56) and l == 16:
        print("MASTERCARD")
    
    # Check VISA
    elif n in range(40, 50) and (l == 13 or l == 16):
        print("VISA")
    
    # Else INVALID
    else:
        print("INVALID")
        
        
if __name__ == "__main__":
    main()